migration/idempotent/018-metric-rollup-maintenance.sql

```diff
@@ -0,0 +1,167 @@
+'Creates a Cagg refresh job that refreshes all Caggs registered by register_metric_view().
+This function creates a refresh job only if no execute_caggs_refresh_policy() exists currently with the given refresh_interval.';
+
+CREATE OR REPLACE PROCEDURE _prom_catalog.execute_caggs_compression_policy(job_id int, config jsonb)
+    SET search_path = pg_catalog, pg_temp
+AS
+$$
+DECLARE
+    r RECORD;
+
+BEGIN
+    FOR r IN
+        SELECT
+            format('%I.%I', m.table_schema, m.table_name) AS compressible_cagg,
+            m.view_refresh_interval AS refresh_interval
+        FROM
+            _prom_catalog.metric m
+        INNER JOIN
+            timescaledb_information.continuous_aggregates cagg
+        ON (m.table_name = cagg.view_name AND m.table_schema = cagg.view_schema)
+        WHERE
+            cagg.compression_enabled AND m.is_view
+    LOOP
+        PERFORM public.compress_chunk(public.show_chunks(r.compressible_cagg, older_than => 100 * r.refresh_interval));
+    END LOOP;
+END;
+$$
+LANGUAGE PLPGSQL;
+COMMENT ON PROCEDURE _prom_catalog.execute_caggs_compression_policy IS 'execute_caggs_compression_policy is responsible to compress Caggs registered via
+register_metric_view() in _prom_catalog.metric. It goes through all the entries in the _prom_catalog.metric and tries to compress any Cagg that supports compression.
+These include metric-rollups and custom Caggs based downsampling.
+Note: execute_caggs_compression_policy runs every X interval and compresses only the inactive chunks of those Caggs which have timescaledb.compress = true.
+By default, these include metric-rollups.';
+
+CREATE OR REPLACE PROCEDURE _prom_catalog.execute_caggs_retention_policy(job_id int, config jsonb)
+    SET search_path = pg_catalog, pg_temp
+AS
+$$
+DECLARE
+    r RECORD;
+    _retention INTERVAL;
+
+BEGIN
+    -- For retention, we have 2 cases:
+    -- Case 1: Retention for metric-rollups
+    -- These Caggs have a defined retention duration that is stored in the _prom_catalog.rollup
+    --
+    -- Case 2: Retention for custom Caggs
+    -- These Caggs do not have a well definied policy for retention. Hence, we will assume the
+    -- retention to be the retention of their parent hypertable.
+
+    -- Case 1.
+    FOR r IN
+        SELECT
+            format('%I.%I', rp.schema_name, m.table_name) AS cagg,
+            rp.retention as rollup_retention
+        FROM _prom_catalog.metric m
+        INNER JOIN _prom_catalog.metric_rollup mr ON (m.id = mr.metric_id)
+        INNER JOIN _prom_catalog.rollup rp ON (mr.rollup_id = rp.id)
+    LOOP
+        PERFORM public.drop_chunks(r.cagg, older_than => r.rollup_retention);
+    END LOOP;
+
+    -- Case 2.
+    FOR r IN
+        SELECT
+            format('%I.%I', table_schema, metric_name) AS cagg,
+            metric_name
+        FROM _prom_catalog.metric WHERE is_view AND NOT is_rollup
+    LOOP
+        _retention := (SELECT _prom_catalog.get_metric_retention_period(r.metric_name));
```

Comment:
Note to self: We should extract the underlying table and do this work, since the r.metric_name here will be a Cagg name, that may not be an actual metric name.